### DISCUSSION TITLE
Integrate GoPro with Tailscale

### DISCUSSION THREAD

**Discussion by compoundvpn** [gopro/labs#827] [Feature Requests]
Hi David, 

Here's a feature request / idea to enhance Labs by creating a secure peer-to-peer mesh network of GoPro cameras using Tailscale accessible from anywhere on earth (or space!) via the Internet.

https://tailscale.com/

Tailscale was founded by ex-Google engineers, perhaps you already know them or have contact via other ventures with Google.

From one of the recent GoPro job listings it states "The firmware of GoPro’s camera is organized in several layers, with various roles and responsibilities. All these layers are running on an RTOS system, including the top-level layer: the GUI. GoPro’s cameras also have a Linux core available". Job listing: https://jobs.gopro.com/jobs/engineering/paris-issy-les-moulineaux/embedded-software-and-imaging-firmware-internship-ui-on-linux/5665730?gh_jid=5665730#/

So it should be a quick win to install Tailscale into the GoPro ecosystem since Tailscale already supports many different versions of Linux.

Here's some ideas, from simple to more complex scenarios with all GoPro cameras running Tailscale and connected to wifi.

1. Consumer usage: GoPro camera powered on with battery or via USB, connected to home wifi. Single user, connect to camera over ssh, copy down files from the sdcard via Taildrop to a local machine or over to a Synology NAS for backup.  Taildrop can be used to transfer files between devices. https://tailscale.com/kb/1106/taildrop

2. Consumer usage / Audobon Society - One or more GoPro cameras out in the forest, powered by solar panel and battery, along with a 4G/5G data modem. Since the user is very far away from the GoPro, they are unable to change settings via QRCode. It's also connected to the Tailscale mesh network, so they can remotely ssh into it, send one or more GoPro Labs commands, and start/stop recording. Copy files over remotely to their NAS or to their AWS instance via scripts.  Once the data has been captured, remotely command the camera to format the sdcard. If the video wasn't the quality they wanted, they could send commands to the GoPro to change the video settings, or change to Night Lapse mode.

4. Imagine a collaboration a few big entities like Disney and NASA/SpaceX. They have hundreds of custom GoPro's with Tailscale installed, each camera with a unique IP address and DNS entry. Live feeds can be streamed over the secure network from some cameras while other cameras are used for data logging and that data is being pulled from the camera over to high-speed computing resources over at GCP or AWS for analysis, all connected with Tailscale. All entities can collaborate together and access the many GoPro devices on the network.

5. Imagine a large trucking company with thousands of trucks on the road, all logging data with the GoPro's, and also acting as dashcams. Through StarLink and over Tailscale, the administrators pull in that day's data at night.

For the regular Labs consumer, it would be a fun way to integrate GoPro's into their secure Tailscale network and find other use cases for GoPro, thereby enhancing that feature of Labs.

For a corporate customer, it provides them with the ability to remotely control a network-connected GoPro from anywhere and manipulate the camera settings, push/pull data, update firmware remotely, etc. All of this over a secure Wireguard connection.

Years ago GoPro had built an array of GoPro cameras. Imagine today, with Labs enabled with Tailscale, you could control and send commands and options to all cameras in the array so long as they are all connected to the network wirelessly.

Tailscale has support for embedded systems and small binaries - https://tailscale.com/kb/1207/small-tailscale

Tailscale for Media Companies - https://tailscale.com/learn/tailscale-for-media-companies

Just a few ideas above :-) 

What do you think?

**Comment by dnewman-gpsw** [comment]
Can you do that same things with Open GoPro?  https://gopro.github.io/OpenGoPro/

For Labs the limitation is it is only running under RTOS, not Linux, so for either Open GoPro or Tailscale that bridge is yet to be built.

**Reply by compoundvpn** [reply]
Yes, same things can be done with OpenGoPro and Labs locally under the same wifi and nearby to activate the camera via BLE. 

Adding Tailscale allows the  Quik app (or custom app) and GoPro camera to be on **separate** networks, and physically far away from each other, and having multiple GoPros on multiple different networks in many different physical locations, all on a secure (WireGuard) mesh network via Tailscale. 

Taking it a bit further for future hardware versions of cameras, or for custom hardware, add in an eSIM to the camera. This would allow the camera to connect to a 4G/5G network, no need for a separate wifi router, and using Tailscale be able to connect to the Quik app anywhere in the world .

There is also potential here for hardware developers to incorporate Wifi/5G routers with BLE inside, running Tailscale. Very similar to existing offers from Gl-iNet whose OpenWRT routers support TailScale as well.  

Consider the following scenario - a camera in the remote location powered by USB solar powerbank. A nearby wireless router with Wifi and 5G and BLE capability. A person physically far away on a 5G network on their phone running Quik and Tailscale on their phone. They can then select the appropriate camera connected and active on the Tailscale mesh network and start / stop recording, send commands, remotely configure it etc.

Incorporating Tailscale into the GoPro would open up a whole new world of possibilities and use-cases for the camera.

**Comment by Espusen** [comment]
This sounds like an amazing potential for growth for LABS!

Can remote reprogramming be done with Open Gopro??

**Reply by dnewman-gpsw** [reply]
> Can remote reprogramming be done with Open Gopro?

Open GoPro should be able to do anything that the Quik app does for the camera control, file transfers, and configurations.
